In [5]:
import sys

In [6]:
sys.path.append(r'C:\Users\jk\Desktop\data science - infoshare\projekt_SQL\jdszr4-edc\2-projekt-python/misc')

In [7]:
from matplotlib import pyplot
import pandas as pd
import numpy as np

In [8]:
import data_preparation as prep

In [9]:
df = prep.df
df

Age  Attrition  BusinessTravel  DailyRate  Department  DistanceFromHome  \
0      41          1               1       1102           2                 1   
1      49          0               2        279           1                 8   
2      37          1               1       1373           1                 2   
3      33          0               2       1392           1                 3   
4      27          0               1        591           1                 2   
...   ...        ...             ...        ...         ...               ...   
1465   36          0               2        884           1                23   
1466   39          0               1        613           1                 6   
1467   27          0               1        155           1                 4   
1468   49          0               2       1023           2                 2   
1469   34          0               1        628           1                 8   

      Education  EducationField  EnvironmentSatisfaction  Gender  ...  \
0             2               1                        2       0  ...   
1             1               1                        3       1  ...   
2             2               5                        4       1  ...   
3             4               1                        4       0  ...   
4             1               3                        1       1  ...   
...         ...             ...                      ...     ...  ...   
1465          2               3                        3       1  ...   
1466          1               3                        4       1  ...   
1467          3               1                        2       1  ...   
1468          3               3                        4       1  ...   
1469          3               3                        2       1  ...   

      PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0                     3                         1                 0   
1                     4                         4                 1   
2                     3                         2                 0   
3                     3                         3                 0   
4                     3                         4                 1   
...                 ...                       ...               ...   
1465                  3                         3                 1   
1466                  3                         1                 1   
1467                  4                         2                 1   
1468                  3                         4                 0   
1469                  3                         1                 0   

      TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
0                     8                      0                1   
1                    10                      3                3   
2                     7                      3                3   
3                     8                      3                3   
4                     6                      3                3   
...                 ...                    ...              ...   
1465                 17                      3                3   
1466                  9                      5                3   
1467                  6                      0                3   
1468                 17                      3                2   
1469                  6                      3                4   

      YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0                  6                   4                        0   
1                 10                   7                        1   
2                  0                   0                        0   
3                  8                   7                        3   
4                  2                   2                        2   
...              ...                 ...                      ...   
1465           

# Analiza poziomu satysfakcji pod względem całkowitej liczby lat doświadczenia - total working years

In [10]:
df_TotalWorkingYears = df[['TotalWorkingYears', 'EnvironmentSatisfaction', 'JobSatisfaction', 'RelationshipSatisfaction']]
df_TotalWorkingYears.sort_values(by=['TotalWorkingYears'])

TotalWorkingYears  EnvironmentSatisfaction  JobSatisfaction  \
1153                  0                        2                4   
1311                  0                        2                3   
457                   0                        2                2   
828                   0                        3                3   
615                   0                        4                4   
...                 ...                      ...              ...   
85                   37                        4                4   
1301                 37                        2                2   
98                   38                        4                3   
595                  40                        4                2   
126                  40                        4                4   

      RelationshipSatisfaction  
1153                         3  
1311                         3  
457                          4  
828                          4  
615                          3  
...                        ...  
85                           1  
1301                         4  
98                           3  
595                          4  
126                          4  

[1470 rows x 4 columns]

In [7]:
#sprawdzenie danych dla TotalWorkingYears

In [8]:
df.describe()['TotalWorkingYears']

count    1470.000000
mean       11.279592
std         7.780782
min         0.000000
25%         6.000000
50%        10.000000
75%        15.000000
max        40.000000
Name: TotalWorkingYears, dtype: float64

In [21]:
df_25 = df['TotalWorkingYears'].quantile(0.25)
df_50 = df['TotalWorkingYears'].quantile(0.5)
df_75 = df['TotalWorkingYears'].quantile(0.75)
df_90 = df['TotalWorkingYears'].quantile(0.9)
print(df_25, df_50, df_75, df_90)

6.0 10.0 15.0 23.0


In [60]:
#na podstawie wyliczonych kwantyli, wyznaczam przedziały, dla których będę przeprowadzać analizę
#T1 - osoby, które pracują 6 lat lub krócej
#T2 - osoby, które pracują dłużej niż 6 lat, ale krócej niż 10
#T3 - osoby, które pracują dłużej niż 10 lat, ale krócej niż 15
#T4 - osoby, które pracują dłużej niż 15 lat, ale krócej niż 23
#T5 - osoby, które pracują dłużej niż 23 lata

In [113]:
T1 = df[(df.TotalWorkingYears <= 6)]
T1_enivronment = T1['EnvironmentSatisfaction'].mean()
T1_job =T1['JobSatisfaction'].mean()
T1_relationship = T1['RelationshipSatisfaction'].mean()

print(T1_enivronment,T1_job, T1_relationship)

2.7165532879818595 2.7301587301587302 2.7346938775510203


In [114]:
T2 = df[(df.TotalWorkingYears > 6) & (df.TotalWorkingYears <= 10)]
T2_enivronment = T2['EnvironmentSatisfaction'].mean()
T2_job = T2['JobSatisfaction'].mean()
T2_relationship = T2['RelationshipSatisfaction'].mean()
print(T2_enivronment,T2_job, T2_relationship)

2.711618257261411 2.7427385892116183 2.6825726141078836


In [115]:
T3 = df[(df.TotalWorkingYears > 10) & (df.TotalWorkingYears <= 15)]
T3_enivronment = T3['EnvironmentSatisfaction'].mean()
T3_job = T3['JobSatisfaction'].mean()
T3_relationship = T3['RelationshipSatisfaction'].mean()
print(T3_enivronment,T3_job, T3_relationship)

2.738219895287958 2.816753926701571 2.5968586387434556


In [116]:
T4 = df[(df.TotalWorkingYears > 15) & (df.TotalWorkingYears <= 23)]
T4_enivronment = T4['EnvironmentSatisfaction'].mean()
T4_job = T4['JobSatisfaction'].mean()
T4_relationship = T4['RelationshipSatisfaction'].mean()
print(T4_enivronment,T4_job, T4_relationship)

2.747787610619469 2.6238938053097347 2.774336283185841


In [117]:
T5 = df[df.TotalWorkingYears > 23]
T5_enivronment = T5['EnvironmentSatisfaction'].mean()
T5_job = T5['JobSatisfaction'].mean()
T5_relationship = T5['RelationshipSatisfaction'].mean()
print(T5_enivronment,T5_job, T5_relationship)

2.707692307692308 2.723076923076923 2.8076923076923075


|_PRZEDZIAŁ CAŁKOWITEGO DOŚWIADCZENIA_ ||_ŚREDNIA SATYSFAKCJA Z PRACY_ ||_ŚREDNIA SATYSFAKCJA ZE ŚRODOWISKA PRACY_||_ŚREDNIA SATYSFAKCJA Z RELACJI W PRACY_|
|-----||-----||-----||-----|
| <6||2.72||2.73||2.73|
|6 - 10||2.71||2.74||2.68|
|10 - 15||2.74||2.82||2.60|
|15 - 23||2.75||2.62||2.77|
|>23||2.71||2.72||2.81|

In [43]:
def corr_color(val):
    if abs(val) == 1.0:
        color = 'grey'
    elif abs(val) > 0.8:
        color = 'red'
    elif abs(val) > 0.5:
        color = 'orange'
    elif abs(val) > 0.3:
        color = 'yellow'
    else:
        color = 'blue'
    return 'color: %s' % color

In [66]:
T1_corr = T1.corr()
T1_corr = T1_corr.style.applymap(corr_color)
T1_corr

Wśród gurpy osób pracujących krócej niż 6 lat, nie ma korelacji pomiędzy stażem pracy a którymkolwiek z badanych czynników satysfakcji.

In [67]:
T2_corr = T2.corr()
T2_corr = T2_corr.style.applymap(corr_color)
T2_corr

Wśród gurpy osób pracujących dłużej niż 6 lat, lecz krócej niż 10, nie ma korelacji pomiędzy stażem pracy a którymkolwiek z badanych czynników satysfakcji

In [68]:
T3_corr = T3.corr()
T3_corr = T3_corr.style.applymap(corr_color)
T3_corr

Wśród gurpy osób pracujących dłużej niż 10 lat, lecz krócej niż 15, nie ma korelacji pomiędzy stażem pracy a którymkolwiek z badanych czynników satysfakcji

In [69]:
T4_corr = T4.corr()
T4_corr = T4_corr.style.applymap(corr_color)
T4_corr

Wśród gurpy osób pracujących dłużej niż 15 lat lecz krócej niż 23 lata, nie ma korelacji pomiędzy stażem pracy a którymkolwiek z badanych czynników satysfakcji

In [70]:
T5_corr = T5.corr()
T5_corr = T5_corr.style.applymap(corr_color)
T5_corr

Wśród gurpy osób pracujących dłużej niż 23 lata, nie ma korelacji pomiędzy stażem pracy a którymkolwiek z badanych czynników satysfakcji

# Analiza poziomu satysfakcji pod względem liczby lat doświadczenia w firmie - years at company

In [53]:
df_YearsAtCompany = df[['YearsAtCompany', 'EnvironmentSatisfaction', 'JobSatisfaction', 'RelationshipSatisfaction']]
df_YearsAtCompany.sort_values(by=['YearsAtCompany'])

YearsAtCompany  EnvironmentSatisfaction  JobSatisfaction  \
1311               0                        2                3   
1371               0                        4                1   
460                0                        1                3   
1068               0                        3                1   
457                0                        2                2   
...              ...                      ...              ...   
561               34                        3                1   
1116              36                        3                4   
270               36                        4                1   
98                37                        4                3   
126               40                        4                4   

      RelationshipSatisfaction  
1311                         3  
1371                         3  
460                          1  
1068                         3  
457                          4  
...                        ...  
561                          1  
1116                         3  
270                          3  
98                           3  
126                          4  

[1470 rows x 4 columns]

In [56]:
df.describe()['YearsAtCompany']

count    1470.000000
mean        7.008163
std         6.126525
min         0.000000
25%         3.000000
50%         5.000000
75%         9.000000
max        40.000000
Name: YearsAtCompany, dtype: float64

In [54]:
dfy_25 = df['YearsAtCompany'].quantile(0.25)
dfy_50 = df['YearsAtCompany'].quantile(0.5)
dfy_75 = df['YearsAtCompany'].quantile(0.75)
dfy_90 = df['YearsAtCompany'].quantile(0.9)
print(dfy_25, dfy_50, dfy_75, dfy_90)

3.0 5.0 9.0 15.0


In [55]:
#na podstawie wyliczonych kwantyli, wyznaczam przedziały, dla których będę przeprowadzać analizę
#Y1 - osoby, które pracują w firmie 3 lata lub krócej
#Y2 - osoby, które pracują w firmie dłużej niż 3 lata, ale krócej niż 5
#Y3 - osoby, które pracują w firmie dłużej niż 5 lat, ale krócej niż 9
#Y4 - osoby, które pracują w firmie dłużej niż 9 lat,a le krócej niż 15
#Y5 - osoby, które pracują w firmie dłużej niż 15 lat

In [118]:
Y1 = df[df.YearsAtCompany <= 3]
Y1_enivronment = Y1['EnvironmentSatisfaction'].mean()
Y1_job =Y1['JobSatisfaction'].mean()
Y1_relationship = Y1['RelationshipSatisfaction'].mean()

print(Y1_enivronment,Y1_job, Y1_relationship)

2.6702127659574466 2.7148936170212767 2.74468085106383


In [119]:
Y2 = df[(df.YearsAtCompany > 3) & (df.YearsAtCompany <= 5)]
Y2_enivronment = Y2['EnvironmentSatisfaction'].mean()
Y2_job =Y2['JobSatisfaction'].mean()
Y2_relationship = Y2['RelationshipSatisfaction'].mean()

print(Y2_enivronment,Y2_job, Y2_relationship)

2.735294117647059 2.689542483660131 2.7058823529411766


In [120]:
Y3 = df[(df.YearsAtCompany > 5) & (df.YearsAtCompany <= 9)]
Y3_enivronment = Y3['EnvironmentSatisfaction'].mean()
Y3_job =Y3['JobSatisfaction'].mean()
Y3_relationship = Y3['RelationshipSatisfaction'].mean()

print(Y3_enivronment,Y3_job, Y3_relationship)

2.8292682926829267 2.7560975609756095 2.652439024390244


In [121]:
Y4 = df[(df.YearsAtCompany > 9) & (df.YearsAtCompany <= 15)]
Y4_enivronment = Y4['EnvironmentSatisfaction'].mean()
Y4_job =Y4['JobSatisfaction'].mean()
Y4_relationship = Y4['RelationshipSatisfaction'].mean()

print(Y4_enivronment,Y4_job, Y4_relationship)

2.6973684210526314 2.8157894736842106 2.675438596491228


In [75]:
Y5 = df[df.YearsAtCompany > 15]
Y5_enivronment = Y4['EnvironmentSatisfaction'].mean()
Y5_job =Y5['JobSatisfaction'].mean()
Y5_relationship = Y5['RelationshipSatisfaction'].mean()

print(Y5_enivronment,Y5_job, Y5_relationship)

2.6971153846153846 2.652173913043478 2.818840579710145


|_PRZEDZIAŁ DOŚWIADCZENIA W FiRMIE_ ||_ŚREDNIA SATYSFAKCJA Z PRACY_ ||_ŚREDNIA SATYSFAKCJA ZE ŚRODOWISKA PRACY_||_ŚREDNIA SATYSFAKCJA Z RELACJI W PRACY_|
|-----||-----||-----||-----|
| <3||2.67||2.71||2.74|
|3 - 5||2.74||2.69||2.71|
|5 - 9||2.83||2.76||2.65|
|9 - 15||2.70||2.82||2.68|
|>15||2.70||2.65||2.82|

In [78]:
Y1_corr = Y1.corr()
Y1_corr = Y1_corr.style.applymap(corr_color)
Y1_corr

Wśród osób pracujących w firmie krócej niż 3 lata nie ma korelacji pomiędzy stażem, a badanymi czynnikami satysfakcji. 

Dodatkowo: silnie dodtanie korelacje pomiędzy stażem w firmie, a stażem na danym stanowisku, czasem od ostatniego awansu oraz ze stażem pracy z obecnym managerem.

In [80]:
Y2_corr = Y2.corr()
Y2_corr = Y2_corr.style.applymap(corr_color)
Y2_corr

In [81]:
Y3_corr = Y3.corr()
Y3_corr = Y3_corr.style.applymap(corr_color)
Y3_corr

Wśród osób pracujących w firmie dłużej niż 5 lat lecz krócej niż 9 nie ma korelacji pomiędzy stażem, a badanymi czynnikami satysfakcji. 

Dodatkowo: niewielkie dodtanie korelacje pomiędzy stażem w firmie, a stażem na danym stanowisku, czasem od ostatniego awansu oraz ze stażem pracy z obecnym managerem.

In [82]:
Y4_corr = Y4.corr()
Y4_corr = Y4_corr.style.applymap(corr_color)
Y4_corr

Wśród osób pracujących w firmie dłużej niż 9 lat lecz krócej niż 15 nie ma korelacji pomiędzy stażem, a badanymi czynnikami satysfakcji. 

Dodatkowo: niewielkie dodtanie korelacje pomiędzy stażem w firmie, a stażem na danym stanowisku i ze stażem pracy z obecnym managerem.

In [83]:
Y5_corr = Y5.corr()
Y5_corr = Y5_corr.style.applymap(corr_color)
Y5_corr

Wśród osób pracujących w firmie dłużej niż 15 lat nie ma korelacji pomiędzy stażem, a badanymi czynnikami satysfakcji. 

Dodatkowo: dodtanie korelacje pomiędzy stażem w firmie, a wiekiem, poziomem pracy (job level), miesięcznym przychodem oraz całkowitym stażem pracy.

# Analiza poziomu satysfakcji pod względem liczby lat doświadczenia na danym stanowisku - years at current role

In [85]:
df_YearsInCurrentRole = df[['YearsInCurrentRole', 'EnvironmentSatisfaction', 'JobSatisfaction', 'RelationshipSatisfaction']]
df_YearsInCurrentRole.sort_values(by=['YearsInCurrentRole'])

YearsInCurrentRole  EnvironmentSatisfaction  JobSatisfaction  \
742                    0                        1                4   
1233                   0                        2                4   
906                    0                        3                3   
1234                   0                        3                2   
258                    0                        3                4   
...                  ...                      ...              ...   
1327                  17                        1                1   
1351                  17                        4                4   
231                   17                        3                4   
123                   18                        1                3   
190                   18                        3                3   

      RelationshipSatisfaction  
742                          2  
1233                         2  
906                          4  
1234                         4  
258                          2  
...                        ...  
1327                         4  
1351                         2  
231                          4  
123                          3  
190                          1  

[1470 rows x 4 columns]

In [86]:
df.describe()['YearsInCurrentRole']

count    1470.000000
mean        4.229252
std         3.623137
min         0.000000
25%         2.000000
50%         3.000000
75%         7.000000
max        18.000000
Name: YearsInCurrentRole, dtype: float64

In [87]:
dfc_25 = df['YearsInCurrentRole'].quantile(0.25)
dfc_50 = df['YearsInCurrentRole'].quantile(0.5)
dfc_75 = df['YearsInCurrentRole'].quantile(0.75)
dfc_90 = df['YearsInCurrentRole'].quantile(0.9)
print(dfc_25, dfc_50, dfc_75, dfc_90)

2.0 3.0 7.0 9.0


In [88]:
#na podstawie wyliczonych kwantyli, wyznaczam przedziały, dla których będę przeprowadzać analizę
#C1 - osoby, które pracują na obecnym stanowisku 2 lata lub krócej
#C2 - osoby, które pracują na obecnym stanowisku dłużej niż 2 lata, ale krócej lub równo 3
#C3 - osoby, które pracują na obecnym stanowisku dłużej niż 3 lata, ale krócej lub równo 7
#C4 - osoby, które pracują na obecnym stanowisku dłużej niż 7 lata, ale krócej lub równo 9
#C5 - osoby, które pracują na obecnym stanowisku dłużej niż 9 lat

In [98]:
C1 = df[df.YearsInCurrentRole <= 2]
C1_enivronment = C1['EnvironmentSatisfaction'].mean()
C1_job = C1['JobSatisfaction'].mean()
C1_relationship = C1['RelationshipSatisfaction'].mean()

print(C1_enivronment,C1_job, C1_relationship)

2.695393759286776 2.736998514115899 2.7726597325408617


In [99]:
C2 = df[(df.YearsInCurrentRole > 2) & (df.YearsInCurrentRole <= 3)]
C2_enivronment = C2['EnvironmentSatisfaction'].mean()
C2_job = C2['JobSatisfaction'].mean()
C2_relationship = C2['RelationshipSatisfaction'].mean()

print(C2_enivronment,C2_job, C2_relationship)

2.688888888888889 2.6222222222222222 2.6222222222222222


In [100]:
C3 = df[(df.YearsInCurrentRole > 3) & (df.YearsInCurrentRole <= 7)]
C3_environment = C3['EnvironmentSatisfaction'].mean()
C3_job = C3['JobSatisfaction'].mean()
C3_relationship = C3['RelationshipSatisfaction'].mean()

print(C3_environment,C3_job, C3_relationship)

2.7343358395989976 2.7593984962406015 2.6340852130325816


In [102]:
C4 = df[(df.YearsInCurrentRole > 7) & (df.YearsInCurrentRole <= 9)]
C4_environment = C3['EnvironmentSatisfaction'].mean()
C4_job = C4['JobSatisfaction'].mean()
C4_relationship = C4['RelationshipSatisfaction'].mean()

print(C4_environment,C4_job, C4_relationship)

2.7343358395989976 2.7051282051282053 2.6858974358974357


In [103]:
C5 = df[df.YearsInCurrentRole > 9]
C5_environment = C5['EnvironmentSatisfaction'].mean()
C5_job = C5['JobSatisfaction'].mean()
C5_relationship = C5['RelationshipSatisfaction'].mean()

print(C5_environment,C5_job, C5_relationship)

2.7570093457943927 2.7289719626168223 2.7757009345794392


|_PRZEDZIAŁ DOŚWIADCZENIA NA DANYM STANOWISKUE_ ||_ŚREDNIA SATYSFAKCJA Z PRACY_ ||_ŚREDNIA SATYSFAKCJA ZE ŚRODOWISKA PRACY_||_ŚREDNIA SATYSFAKCJA Z RELACJI W PRACY_|
|-----||-----||-----||-----|
| <2||2.70||2.73||2.77|
|2 - 3||2.69||2.62||2.62|
|3 - 5||2.73||2.76||2.63|
|5 - 9||2.73||2.73||2.69|
|>9||2.76||2.73||2.78|

In [104]:
C1_corr = C1.corr()
C1_corr = C1_corr.style.applymap(corr_color)
C1_corr

Wśród osób o stażu nn obecnym stanowisku krótszym niż 2 lata nie znaleziono korelacji pomiędzy stażem na obecnym stanowisku a badanymi czynnikami satysfakcji.

Dodatkowo: znaleziono korelację pomiędzy stażem na obecnym stanowisku a stażem z obecnym managerem.

In [107]:
C2_corr = C2.corr()
C2_corr = C2_corr.style.applymap(corr_color)
C2_corr

In [108]:
#Wśród osób o stażu nn obecnym stanowisku krótszym lub równym 3 lata nie znaleziono korelacji pomiędzy stażem na obecnym stanowisku a badanymi czynnikami satysfakcji.

#Dodatkowo: znaleziono korelację pomiędzy stażem na obecnym stanowisku a stażem z obecnym managerem.

In [109]:
C3_corr = C3.corr()
C3_corr = C3_corr.style.applymap(corr_color)
C3_corr

Wśród osób o stażu nn obecnym stanowisku dłuższym niż 3 lata lecz krótszym lub równym 5 lat nie znaleziono korelacji pomiędzy stażem na obecnym stanowisku a badanymi czynnikami satysfakcji.

Dodatkowo: znaleziono korelację pomiędzy stażem na obecnym stanowisku a stażem z obecnym managerem oraz stażem w firmie.

In [110]:
C4_corr = C4.corr()
C4_corr = C4_corr.style.applymap(corr_color)
C4_corr

Wśród osób o stażu nn obecnym stanowisku dłuższym niż 5 lata lecz krótszym lub równym 9 lat nie znaleziono korelacji pomiędzy stażem na obecnym stanowisku a badanymi czynnikami satysfakcji.

In [111]:
C5_corr = C5.corr()
C5_corr = C5_corr.style.applymap(corr_color)
C5_corr

Wśród osób o stażu nn obecnym stanowisku dłuższym niż 9 lat nie znaleziono korelacji pomiędzy stażem na obecnym stanowisku a badanymi czynnikami satysfakcji.

Dodatkowo: znaleziono korelację pomiędzy stażem na obecnym stanowisku a miesięcznym przychodem i poziomem stanowiska.